Filter low gas txns early

benaadams opened the pull request:
Resolves #4793

Updated #5158 rebased on `Refactor tx filtering` #5281 (main branch)

## Changes

* Changes `FeeTooLowFilter` for `BalanceTooLowFilter`
* Early accept on local or non-full txn pool prior to checking state
* Introduces `FeeTooLowFilter` early which just considers gas fee without accessing state
* Reject transactions with zero balance before checking txns
* Introduce new metric `PendingTransactionsUnresolvableSender`
* Introduce new metric `PendingTransactionsGasLimitTooHigh`
* Introduce new metric `Metrics.PendingTransactionsZeroBalance`
* Introduce new metric `PendingTransactionsTooLowBalance`
* Introduce new metric `PendingTransactionsPassedFiltersButCannotCompeteOnFees`
* Introduce new metric `PendingTransactionsMalformed`
* Introduce new metric `PendingTransactionsLowNonce`
* Add info report on txPool behind config
  `  "TxPool": {
    "ReportMinutes": 1
  }` 

## Types of changes

Stronger tx filtering
```
Txn Pool State (2,048 txns queued)
------------------------------------------------
Sent
* Transactions:                                0
* Hashes:                             12,361,372
------------------------------------------------
Total Received:                          587,284
------------------------------------------------
Discarded at Filter Stage:
1.  GasLimitTooHigh:                           2
2.  Too Low Fee:                         215,506
3.  Malformed                              1,927
4.  Duplicate:                           144,440
5.  Unknown Sender:                            0
6.  Zero Balance:                         14,598
7.  Balance < tx.value:                   27,516
8.  Nonce used:                           45,532
9.  Nonces skipped:                       40,029
10. Balance Too Low:                      20,674
11. Cannot Compete:                          523
------------------------------------------------
Validated via State:                     225,409
------------------------------------------------
Total Discarded:                         510,224
------------------------------------------------
Discard Ratios:
* Pre-state Discards:                  70.54706%
* Received Discarded:                  86.87858%
------------------------------------------------
Total Added:                              76,537
* Eip1559 Added:                          46,851
------------------------------------------------
Total Evicted:                            49,652
------------------------------------------------
Ratios:
* Eip1559 Transactions:                88.88889%
* DarkPool Level1:                      1.48148%
* DarkPool Level2:                      1.48148%
------------------------------------------------
```

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes - changed existing tests
- [ ] No

marcindsobczak reviewed (approved):
`BalanceTooLowFilter` is redundant now.

After recent changes in `TooExpensiveTxFilter` we are filtering out all txs which balance is not able to pay `MaxFeePerGas`. It means, that we don't need to check how much fee can pay tx from sender with balance X - now all txs can always pay full `MaxFeePerGas`. So right now `BalanceTooLowFilter` is not needed - txs with TooLowFee was already filtered in `GasTooLowFilter`, and balance would be checked in `TooExpensiveTxFilter`.

So I think we can remove `BalanceTooLowFilter`.
Then `AcceptTxResult InsufficientBalanceToCompete` and metric `PendingTransactionsBalanceToLowToCompeteOnFee` are redundant too.

And `GasTooLowFilter` can be named `FeeTooLowFilter` as before to avoid confusion (GasTooLowFilter seems like something checking gas limit, not GasPrice/MaxFeePerGas).

And actually then we could rename `TooExpensiveTxFilter` to more relevant `BalanceTooLowFilter` 😄

WDYT @LukaszRozmej @MarekM25 ?
I think it's mergable in current shape. We can merge it and I can push few refactors in separate PR to not bother @benaadams about it

benaadams commented:
> BalanceTooLowFilter is redundant now.

Explains why it was hard to trigger in tests 🤣

benaadams commented:
> `BalanceTooLowFilter` is redundant now.

Made the changes.

Also moved `FeeTooLowFilter` before `AlreadyKnownTxFilter`. They are both lightweight, but already know has the additional overhead of allocating for first observation, so reject first if fee too low?

benaadams commented:
@marcindsobczak added some extra `Metrics` 

* Added `Metrics.PendingTransactionsUnresolvableSender` as comment in `UnknownSenderFilter` refers to it being a problem, but wasn't capturing the number
* Added `Metrics.PendingTransactionsGasLimitTooHigh` as always seems to be a debate going on about block max gas
* Added `Metrics.PendingTransactionsTooLowBalance`
* Added `Metrics.PendingTransactionsPassedFiltersButCannotCompeteOnFees` as change to the Txn pool itself dropping the txns to (very last step) to differentiate from earlier step

benaadams commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,72 @@ namespace Nethermind.TxPool.Filters
         public AcceptTxResult Accept(Transaction tx, TxFilteringState state, TxHandlingOptions handlingOptions)
         {
+            bool isTrace = _logger.IsTrace;
+            if (tx.GasLimit <= 0 || GasLimitToWei(tx.GasLimit) < Transaction.BaseTxGasCost * tx.GasPrice)
```
Is gas limit in Gwei and gas price in Wei?

benaadams replied:
This was error; shouldn't have been multiplying by price

benaadams commented:
@marcindsobczak

If the `FeeTooLowFilter` is correct some of the tests will need to change as they are getting rejected due to fee being too low 😅

However would like to validate its correct first before changing them as its quite fierce and rejecting a lot of txns 👀
```
Txn Pool State (815 txns queued)
------------------------------------------
Sent
* Transactions:                          0
* Hashes:                           650838
------------------------------------------
Total Received:                     222369
------------------------------------------
Discarded at Filter Stage:
1.  Too Low Fee:                     88278  <----
2.  Already Known:                   33231
3.  Malformed                           24
4.  GasLimitTooHigh:                     0
5.  Unknown Sender:                      0
6.  Nonce used:                          0
7.  Nonces skipped:                      0
8.  Zero Balance:                   100020
9.  Balance Too Low:                     1
10. Cannot Compete:                      0
------------------------------------------
Total Discarded:                    221554
------------------------------------------
Validation via Storage:             100836
Total Added:                           815
* Eip1559 Added:                       812
------------------------------------------
Total Evicted:                           0
------------------------------------------
Ratios:
* Eip1559 Transactions:              0.00%
* DarkPool Level1:                   0.00%
* DarkPool Level2:                   0.00%
------------------------------------------
```

benaadams commented:
> Txn added seems kinda low? 🤔

Is much less exiting with the gas limit fix 😢

Running for a while to get data

Scooletz commented:
@benaadams Would you consider splitting the stages of the filtering pipeline in a way similar to this proposed at https://github.com/NethermindEth/nethermind/commit/dfb585eec70db33c839b227a834be4945bb0e6e7. With this, in the future one could forget about ordering on its own and implement a specific interfaces. WDYT?

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/BalanceZeroFilter.cs:
```diff
@@ -0,0 +1,36 @@
+            Account account = state.SenderAccount;
+            UInt256 balance = account.Balance;
+
+            if (balance.IsZero)
+            {
+                Metrics.PendingTransactionsZeroBalance++;
+                return AcceptTxResult.InsufficientFunds.WithMessage("Account balance: 0");
+            }
```
There are free transactions that we need to still add in AuRa for example - Malicious/Benin transactions.

benaadams replied:
Added `tx.IsFree()` test

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -550,5 +551,59 @@ private static void AddNodeInfoEntryForTxPool()
+            if (float.IsNaN(preStateDiscards)) preStateDiscards = 0;
+            if (float.IsNaN(receivedDiscarded)) receivedDiscarded = 0;
+
+            logger.Info(@$"Txn Pool State ({Metrics.TransactionCount:0} txns queued)
```
Should be Debug/Trace

benaadams replied:
Is like the JSON RPC report?

benaadams replied:
Put report behind config
```
  "TxPool": {
    "ReportMinutes": 1
  }
```

LukaszRozmej commented on src/Nethermind/Nethermind.Synchronization/Reporting/SyncReport.cs:
```diff
@@ -91,6 +91,7 @@ private void TimerOnElapsed(object? sender, EventArgs e)
             if (_reportId % SyncFullPeersReportFrequency == 0)
             {
                 _syncPeersReport.WriteFullReport();
+                Nethermind.TxPool.TxPool.WriteTxnPoolReport(_logger);
```
Not sure if this should be invoked from SyncReport.

benaadams replied:
Moved to txPool

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,72 @@ namespace Nethermind.TxPool.Filters
+        private readonly static UInt256 GweiToWeiFactor = new UInt256(1_000_000_000);
+        private UInt256 GasLimitToWei(long gasLimitGwei)
+        {
+            return GweiToWeiFactor * new UInt256((ulong)gasLimitGwei);
```
why creating new UInt256?

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,72 @@ namespace Nethermind.TxPool.Filters
         public AcceptTxResult Accept(Transaction tx, TxFilteringState state, TxHandlingOptions handlingOptions)
         {
+            bool isTrace = _logger.IsTrace;
+            if (tx.GasLimit <= 0 || tx.GasLimit < Transaction.BaseTxGasCost)
+            {
+                // Not high enough GasLimit to run a txn
+                Metrics.PendingTransactionsTooLowFee++;
+                if (isTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low gas limit for a txn at price {tx.GasPrice}");
+                return AcceptTxResult.FeeTooLow.WithMessage("GasLimit too low to start txn at GasPrice");
+            }
```
This should be checked already in `MalformedTxFilter`

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -95,14 +94,15 @@ public class TxPool : ITxPool, IDisposable
             _headInfo.HeadChanged += OnHeadChange;
 
             _filterPipeline.Add(new NullHashTxFilter());
+            _filterPipeline.Add(new FeeTooLowFilter(_headInfo, _transactions, _logger));
```
Put it after `MalformedTxFilter` or `GasLimitTxFilter` at the earliest.

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,72 @@ namespace Nethermind.TxPool.Filters
-            UInt256 balance = state.SenderAccount.Balance;
-            UInt256 affordableGasPrice = tx.CalculateAffordableGasPrice(spec.IsEip1559Enabled, _headInfo.CurrentBaseFee, balance);
-            bool isNotLocal = (handlingOptions & TxHandlingOptions.PersistentBroadcast) != TxHandlingOptions.PersistentBroadcast;
+            bool isEip1559Enabled = spec.IsEip1559Enabled;
+            if (isEip1559Enabled && tx.IsEip1559)
+            {
+                if (tx.GasLimit < spec.MinGasLimit || tx.GasPrice < spec.Eip1559BaseFeeMinValue)
+                {
+                    // Amounts too low for spec
+                    Metrics.PendingTransactionsTooLowFee++;
+                    return AcceptTxResult.FeeTooLow.WithMessage("GasLimit or GasPrice too low for Eip1559 spec");
+                }
+            }
```
`MinGasLimit` & `Eip1559BaseFeeMinValue` are block parameters, not transaction ones. Please remove this part.

benaadams replied:
Removed

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,72 @@ namespace Nethermind.TxPool.Filters
+
+            UInt256 affordableGasPrice = tx.CalculateGasPrice(isEip1559Enabled, _headInfo.CurrentBaseFee);
+
+            // Don't accept zero fee txns even if pool is empty as will never run
+            if (affordableGasPrice.IsZero && !_headInfo.CurrentBaseFee.IsZero)
+            {
+                Metrics.PendingTransactionsTooLowFee++;
+                if (isTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, too low payable gas price with options {handlingOptions} from {new StackTrace()}");
+                return AcceptTxResult.FeeTooLow.WithMessage("Affordable FeePerGas of 0 rejected.");
+            }
```
Not sure about AuRa

benaadams replied:
Early accepted if `tx.IsFree()`

benaadams commented:
> @benaadams Would you consider splitting the stages of the filtering pipeline in a way similar to this proposed at [dfb585e](https://github.com/NethermindEth/nethermind/commit/dfb585eec70db33c839b227a834be4945bb0e6e7). With this, in the future one could forget about ordering on its own and implement a specific interfaces. WDYT?

idk, is a fixed pipeline with different costs per filter so might be better to specify it explicitly and have the Jit turn everything into direct calls?

Scooletz commented:
> > @benaadams Would you consider splitting the stages of the filtering pipeline in a way similar to this proposed at [dfb585e](https://github.com/NethermindEth/nethermind/commit/dfb585eec70db33c839b227a834be4945bb0e6e7). With this, in the future one could forget about ordering on its own and implement a specific interfaces. WDYT?
> 
> idk, is a fixed pipeline with different costs per filter so might be better to specify it explicitly and have the Jit turn everything into direct calls?

I meant it only from the discovery and code structuring point of view. If I see two interfaces, I can make an easier decision whether use one or the other. It can be a comment as well in the pipeline, that first should go filter with no state access. Dunno.

benaadams commented:
Looks like it works well
```
Txn Pool State (2012 txns queued)
------------------------------------------
Sent
* Transactions:                          0
* Hashes:                          6032196
------------------------------------------
Total Received:                     282836
------------------------------------------
Discarded at Filter Stage:
1.  GasLimitTooHigh:                     0
2.  Malformed                            3
3.  Too Low Fee:                    102651
4.  Duplicate:                       58567
5.  Unknown Sender:                      0
6.  Zero Balance:                     6874
7.  Balance < tx.value:              13983
8.  Nonce used:                      22531
9.  Nonces skipped:                  26719
10. Balance Too Low:                 10626
11. Cannot Compete:                    869
------------------------------------------
Validated via State:                121615
------------------------------------------
Total Discarded:                    241954
Ratios:
* Pre-state Discards:            66.63167%
* Received Discarded:            85.54569%
------------------------------------------
Total Added:                         40013
* Eip1559 Added:                     25003
------------------------------------------
Total Evicted:                       25729
------------------------------------------
Ratios:
* Eip1559 Transactions:          88.59649%
* DarkPool Level1:                2.63158%
* DarkPool Level2:                2.63158%
------------------------------------------
```
Filters are arranged in costs (though some earlier ones have a higher cost where any in that position would have same additional cost, e.g. initialising data from state)

![image](https://user-images.githubusercontent.com/1142958/222633801-7bcf25d4-42cd-4966-9128-25c4e5e3b60c.png)

benaadams commented:
@marcindsobczak dropping about 86% pre-state access

benaadams commented:
Logic of filter ordering

1.  GasLimitTooHigh - very simple check
2.  Malformed - more complex maths check
3.  Too Low Fee - requires lock to check last txn, above two can work in full parallel
4.  Duplicate - generates Keccak (expensive) and allocates to cache the hash
5.  Unknown Sender - first state access, simple check
6.  Zero Balance - heavier state access, simple check
7.  Balance < tx.value - same filter as above
8.  Nonce used - simple check (using above state)
9.  Nonces skipped - iterates txn pool
10. Balance Too Low - iterates txn pool, more complex maths check
11. Cannot Compete - rejected on add

LukaszRozmej commented on src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V62/Eth62ProtocolHandler.cs:
```diff
@@ -236,7 +239,48 @@ private void Handle(StatusMessage status)
+            bool isTrace = Logger.IsTrace;
+            for (int i = 0; i < transactions.Length; i++)
+            {
+                Transaction tx = transactions[i];
+                tx.DeliveredBy = Node.Id;
+                tx.Timestamp = _timestamper.UnixTime.Seconds;
+                AcceptTxResult accepted = _txPool.SubmitTx(tx, TxHandlingOptions.None);
+                _floodController.Report(accepted);
+
+                if (isTrace) Log(tx, accepted);
+            }
+
+            void Log(Transaction tx, AcceptTxResult accepted)
+            {
+                Logger.Trace($"{Node:c} sent {tx.Hash} tx and it was {accepted} (chain ID = {tx.Signature?.ChainId})");
+            }
```
Code duplication between HandleFast and Handle slow

benaadams replied:
Resolved

LukaszRozmej commented on src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs:
```diff
@@ -12,6 +12,8 @@ namespace Nethermind.Serialization.Rlp
 {
     public class TxDecoder : TxDecoder<Transaction>, ITransactionSizeCalculator
     {
+        public const int MaxDelayedHashTxnSize = 32768;
```
Any particular logic about picking this value? Apart it being power of 2, so works well with ArrayPool buckets.

benaadams replied:
Contract size is 24kB and rounded up; shouldn't be anything larger but good to have cap

benaadams replied:
Happy to set it lower, but KeccakF is expensive

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/FeeTooLowFilter.cs:
```diff
@@ -13,36 +13,46 @@ namespace Nethermind.TxPool.Filters
-            if (isNotLocal
-                && _txs.IsFull()
-                && _txs.TryGetLast(out Transaction? lastTx)
+            if (_txs.IsFull() && _txs.TryGetLast(out Transaction? lastTx)
```
I think we could have a lockless version of TryGetLast, just when something is added to that Set get the last value and save it to some field.

benaadams replied:
`_txs.IsFull()` also takes a lock

LukaszRozmej replied:
Added PR for that: https://github.com/NethermindEth/nethermind/pull/5360 can you try it?

benaadams replied:
Merged

benaadams replied:
> Added PR for that: #5360 can you try it?

Makes `FeeTooLowFilter` cheaper than `MalformedTxFilter` so switched the order

benaadams replied:
<img width="575" alt="image" src="https://user-images.githubusercontent.com/1142958/222701690-56e3f715-0c74-4194-b2d8-6a89de4aae2a.png">

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/ITxPoolConfig.cs:
```diff
@@ -20,6 +20,11 @@ public interface ITxPoolConfig : IConfig
+
         long? GasLimit { get; set; }
+
+        [ConfigItem(DefaultValue = "null",
+            Description = "Minutes between reporting on current state of tx pool.")]
+        int? ReportMinutes { get; set; }
```
Do we need it configurable? We should have this in the actual metrics. It is fine for it to be in some Debug/Trace logs but other than that looks like overkill.

At minimum I would hide it from docs, as this has no use for users.

benaadams replied:
People like them metrics 😅 I'll probably have it on all the time and currently logging in Nethermind isn't typed so you'd have to switch on a tsunami of output and then not be able to see it. So this is a way to switch it on independently.

It is useful as it tells you about the general state of the mem pool.

LukaszRozmej replied:
> Looks good on costs with #5360 (merged into this branch) with 70% of txns being dropped in now light filters before it gets to heavier ones
> 
> <img alt="image" width="574" src="https://user-images.githubusercontent.com/1142958/222710806-422c493a-6fd3-40d2-bed1-4ba8c7f58cb9.png">

one more thing about TxPool filters - would caching balance an nonce somehow help there? Would be good to have some kind of histogram of incoming transactions senders between block (how many have unique senders vs how many senders are duplicated)

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -268,20 +282,85 @@ public AcceptTxResult SubmitTx(Transaction tx, TxHandlingOptions handlingOptions
 
+        private AcceptTxResult FilterTransactions(Transaction tx, TxHandlingOptions handlingOptions, TxFilteringState state)
+        {
+            AcceptTxResult
+
+            // Simple comparison
+            accepted = _filterGasLimitTx.Accept(tx, state, handlingOptions);
```
why 4 lines?

benaadams replied:
Need to release the pooled array if it fails before `NullHashTxFilter` when it runs the `Keccak` generates the hash and releases it

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -268,20 +282,85 @@ public AcceptTxResult SubmitTx(Transaction tx, TxHandlingOptions handlingOptions
             return AddCore(tx, startBroadcast);
         }
 
+        private AcceptTxResult FilterTransactions(Transaction tx, TxHandlingOptions handlingOptions, TxFilteringState state)
```
Is it worth it compared to a clean pipeline? I would say the filters have such a big impact in their logic that this is micro-optimization that doesn't make a dent in the performance, but make the code uglier.

benaadams replied:
Changed

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -550,5 +630,69 @@ private static void AddNodeInfoEntryForTxPool()
+
+        private static void WriteTxnPoolReport(ILogger logger)
+        {
+            if (!logger.IsInfo)
+            {
+                return;
+            }
```
This shouldn't be Info, Debug at max

benaadams replied:
Is an at a glance summary report like Rpc report, peers report; and it gives insight into the state of the mem pool. 

Have you switched on debug logging? 😱
There's no way you can read anything when logging is debug but would have to go through it in a file

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/TxPool.cs:
```diff
@@ -94,20 +103,27 @@ public class TxPool : ITxPool, IDisposable
+            _filterDeployedCode = new DeployedCodeFilter(_specProvider, _accounts);
+
+            int? reportMinutes = txPoolConfig.ReportMinutes;
+            if (reportMinutes.HasValue)
             {
-                _filterPipeline.Add(incomingTxFilter);
+                _timer = TimerFactory.Default.CreateTimer(TimeSpan.FromMinutes(reportMinutes.Value));
+                _timer.AutoReset = false;
+                _timer.Elapsed += TimerOnElapsed;
+                _timer.Start();
```
If the log level isn't enough, don't start the timer.

benaadams replied:
Done

LukaszRozmej reviewed (changes requested):
I am thinking about one more potential filter SortedPool has (protected) method CanInsert, maybe there should be a filter using this method - not sure if worth it in the end (how many transactions would be filtered out there).

benaadams commented:
```
Txn Pool State (2,048 txns queued)
------------------------------------------------
Sent
* Transactions:                                0
* Hashes:                              9,260,030
------------------------------------------------
Total Received:                          579,059
------------------------------------------------
Discarded at Filter Stage:
1.  GasLimitTooHigh:                           0
2.  Too Low Fee:                         198,736
3.  Malformed                                 50
4.  Duplicate:                           123,807
5.  Unknown Sender:                            0
6.  Zero Balance:                         22,516
7.  Balance < tx.value:                   49,316
8.  Nonce used:                           65,766
9.  Nonces skipped:                       35,830
10. Balance Too Low:                      24,567
11. Cannot Compete:                          510
------------------------------------------------
Validated via State:                     256,466
------------------------------------------------
Total Discarded:                         520,588
------------------------------------------------
Discard Ratios:
* Pre-state Discards:                  61.95744%
* Received Discarded:                  89.90241%
------------------------------------------------
Total Added:                              57,961
* Eip1559 Added:                          36,157
------------------------------------------------
Total Evicted:                            37,952
------------------------------------------------
Ratios:
* Eip1559 Transactions:                91.04478%
* DarkPool Level1:                      3.73134%
* DarkPool Level2:                      3.73134%
------------------------------------------------
```

benaadams commented:
Looks good on costs with https://github.com/NethermindEth/nethermind/pull/5360 (merged into this branch) with 62% of txns being dropped in now light filters before it gets to heavier ones

<img width="574" alt="image" src="https://user-images.githubusercontent.com/1142958/222710806-422c493a-6fd3-40d2-bed1-4ba8c7f58cb9.png">

benaadams commented:
4M txs on
```
Txn Pool State (2,048 txns queued)
------------------------------------------------
Sent
* Transactions:                                0
* Hashes:                            101,298,065
------------------------------------------------
Total Received:                        4,925,133
------------------------------------------------
Discarded at Filter Stage:
1.  GasLimitTooHigh:                           5
2.  Too Low Fee:                       1,954,509
3.  Malformed                              2,030
4.  Duplicate:                         2,163,871
5.  Unknown Sender:                            0
6.  Zero Balance:                         37,434
7.  Balance < tx.value:                   61,312
8.  Nonce used:                           89,072
9.  Nonces skipped:                      181,438
10. Balance Too Low:                      44,221
11. Cannot Compete:                        4,796
------------------------------------------------
Validated via State:                     804,718
------------------------------------------------
Total Discarded:                       4,533,892
------------------------------------------------
Discard Ratios:
* Pre-state Discards:                  90.83554%
* Received Discarded:                  92.05623%
------------------------------------------------
Total Added:                             386,445
* Eip1559 Added:                         236,485
------------------------------------------------
Total Evicted:                           255,414
------------------------------------------------
Ratios:
* Eip1559 Transactions:                92.50000%
* DarkPool Level1:                      8.75000%
* DarkPool Level2:                      8.75000%
------------------------------------------------
```

benaadams commented:
Avoided a bunch of allocations also

LukaszRozmej commented on src/Nethermind/Nethermind.TxPool/Filters/BalanceTooLowFilter.cs:
```diff
@@ -62,9 +69,17 @@ public AcceptTxResult Accept(Transaction tx, TxFilteringState state, TxHandlingO
+                {
                     _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, insufficient funds.");
-                return AcceptTxResult.InsufficientFunds.WithMessage($"Account balance: {balance}, cumulative cost: {cumulativeCost}");
+                    return AcceptTxResult.InsufficientFunds.WithMessage($"Account balance: {balance}, cumulative cost: {cumulativeCost}");
```
Those messages are useful for local transactions as they will be returned to the user. We can allocate them only for local ones though. This applies to all filters.

benaadams replied:
Changed to output the good messages for local ones
